CI: Unified build scripts for all platforms

PatTheMav opened the pull request:
### Description
This PR unifies all current build scripts into a single workflow to create required dependencies for macOS and Windows (and lays groundwork to provide pre-compiled FFmpeg libraries for Linux).

It also moves the definition of a dependency into each library's build script file, which allows changes to happen in a single place, independent of the environment (CI vs local) in which a build happens.

All downloads are either confirmed using a checksum or use a specific git commit hash. All functionality is implement for Unix-based systems as a zsh-based framework (due to zsh being the default shell on macOS with bash not being updated anymore) as well as a Powershell-based framework for Windows.

The Powershell-based framework is designed to look and behave as similarly to the zsh-based framework as possible, but follows common Powershell coding styles.

The frameworks allow building specific dependencies (single and multiple) as well as "complete" packages (FFmpeg, Qt) and can be configured to build static or shared variants with common CMake-based build configurations (Debug/Release/RelWithDebInfo/MinSizeRel) translated in a best-effort manner to autoconf configurations.

### Motivation and Context
With the addition of Windows dependencies to the repository, all existing dependency build scripts became publicly documented for the first time, opening the door to sync and combine settings across supported platforms.

As FFmpeg libraries are cross compiled on Linux, this opened the door to have a single build script architecture that is capable to build FFmpeg and its libraries for different architectures while applying the same patches and using the same settings across all systems.

By combining all build scripts into a single system, obs-deps can now be built and released via a single workflow.

### How Has This Been Tested?
* Deps built and released on my own fork
* OBS built on Windows and macOS with dependencies created on CI
* Further testing/dogfooding by other team members would be appreciated

### Types of changes
- New feature (non-breaking change which adds functionality)
- Performance enhancement (non-breaking change which improves efficiency)
- Code cleanup (non-breaking change which makes code smaller or more readable)
- Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented on utils.zsh/functions/progress:
```diff
@@ -0,0 +1,19 @@
+local -a command=($@)
+
+if (( ${+CI} == 0 && _loglevel == 1 )) {
+  local spin='⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏'
```
It truly doesn't matter in the grand scheme of thigns, but sometimes there are three dots per glyph and sometimes there are four. Is that unavoidable?

PatTheMav replied:
I don't think so, the animation is built on these glyphs as-is and makes it look more fluid that way.

RytoEX replied:
Technically, one could always use four dots:
```
⠏⠛⠹⠼⠶⠧
```
Or three dots:
```
⠋⠙⠸⠴⠦⠇
```
But both result in shorter animation loops.
```

RytoEX commented on utils.zsh/functions/set_loglevel:
```diff
@@ -0,0 +1,17 @@
+Set log level, following levels are supported: 0 (quiet), 1 (normal), 2 (verbose), 3 (debug)"
+
+if (( ! # )); then
+  log_error 'Called without arguments.%f'
```
Do single quotes work here?

PatTheMav replied:
Yeah `%f` is a prompt expansion, not a variable expansion.

RytoEX reviewed (changes requested):
I've finally gone over this, and made a lot of notes. Feel free to let me know if there are any items you'd like to defer to a later follow-up.

For libvpx and libsrt, can we implement something to use custom filenames so that both the downloaded files/directories and the checksum files have more descriptive names?

I noticed that `CI/patches/QTBUG-74606.patch` was moved/renamed to `deps.macos/patches/Qt/0001-QTBUG-74606.patch`. We use this patch for Windows too, even though we don't currently build Windows Qt on CI (I plan to address that after this lands). Would the expectation be to put this in a Windows-specific `deps.windows/patches` folder? Or do we have a platform-agnostic patch directory available with this new system?

If possible, I would prefer that we keep all dependency versions the same as they are currently for this PR, unless there are problems with those dependencies with this build system. That way, we can cleanly separate and more easily track these build system changes from the dependency updates and make them easier to find in commit history. I've noted a few places in specific files where I think dependency updates were made.

I noticed a `v1.10.0.tar.gz.sha256` file, but we don't use anything at that version, as far as I know. Was it for libvpx v1.10.0? Though, in line with my previous paragraph, I'd prefer if we kept all of the dependency versions the same in a large workflow change like this, so that we can more easily keep track of actual dependency changes.

The mbedTLS Windows native build is disabled. Is that related to the change to build FFmpeg components as static?

The nasm Windows native build is disabled. Although we do plan to remove this, we currently ship it. I'd prefer this PR stuck to workflow changes and leave dependency version changes and shipping changes for a separate change.

What is the minimum PowerShell version supported? As the GitHub Actions images for both Windows 2019 and Windows 2022 have PowerShell 7.2, and that is relatively easy to install on Windows 10, I'd be fine setting that as a required version if that makes things simpler.

While reviewing this PR, I had initially suggested a lot of string changes for capitalization/case reasons. However, I did notice after that the strings are more or less consistent with what is currently in obs-deps, except maybe SpeexDSP, so I've dropped those from this review for now. I'm just making a note of it here as food for thought for a later date, since I don't want to further complicate this PR with _more_ string nitpicks than I have already included.

Most of the Windows native build scripts do not have their own Switch Parameters. I think it would be useful if they did so that if they are invoked directly, the Switch Parameters should be hinted on the CLI.

This PR currently only builds ajantv2.lib in Release configuration. We currently build both Release and Debug builds of this library.

PatTheMav commented:
> For libvpx and libsrt, can we implement something to use custom filenames so that both the downloaded files/directories and the checksum files have more descriptive names?

Would add unnecessary code complexity, I deliberately did not implement code for that as it makes file name composition (download artifact, extraction folder, checksum file) quite a bit easier.

> I noticed that `CI/patches/QTBUG-74606.patch` was moved/renamed to `deps.macos/patches/Qt/0001-QTBUG-74606.patch`. We use this patch for Windows too, even though we don't currently build Windows Qt on CI (I plan to address that after this lands). Would the expectation be to put this in a Windows-specific `deps.windows/patches` folder? Or do we have a platform-agnostic patch directory available with this new system?

That depends on how we build Qt for Windows. Ideally we'd have a platform-agnostic `deps.qt` folder that contains the patches. We can probably throw both Powershell as well as Zsh files in there without issue, so that's where all the Qt-related code would migrate to then.

> If possible, I would prefer that we keep all dependency versions the same as they are currently for this PR, unless there are problems with those dependencies with this build system. That way, we can cleanly separate and more easily track these build system changes from the dependency updates and make them easier to find in commit history. I've noted a few places in specific files where I think dependency updates were made.

That shouldn't be an issue architecture-wise, I just prefer _in general_ to keep our macOS deps on the most recent version supported by Homebrew, as that community does the heavy lifting of testing and fixing build issues for us. That approach (at least so far) has resulted in the least amount of headaches while at the same time allowing very regular updates on macOS compared to Windows.

> I noticed a `v1.10.0.tar.gz.sha256` file, but we don't use anything at that version, as far as I know. Was it for libvpx v1.10.0? Though, in line with my previous paragraph, I'd prefer if we kept all of the dependency versions the same in a large workflow change like this, so that we can more easily keep track of actual dependency changes.

That's probably because of different versions for Windows and macOS when I originally started the rework and I updated the actual build versions as you introduced changes to Windows deps.

> The mbedTLS Windows native build is disabled. Is that related to the change to build FFmpeg components as static?

That's related to the mbedTLS version build as part of the FFmpeg build is shared by default and didn't have any issues in OBS on Windows in my testing (including RTMPs streaming to Youtube), so no need to build it twice.

The nasm Windows native build is disabled. Although we do plan to remove this, we currently ship it. I'd prefer this PR stuck to workflow changes and leave dependency version changes and shipping changes for a separate change.

That's just a filename change, easy to fix.

> What is the minimum PowerShell version supported? As the GitHub Actions images for both Windows 2019 and Windows 2022 have PowerShell 7.2, and that is relatively easy to install on Windows 10, I'd be fine setting that as a required version if that makes things simpler.

I'd guess the earliest PowerShell Core version (pwsh), I'd guess going with PowerShell 7.2 would be easiest straight out the door.

> While reviewing this PR, I had initially suggested a lot of string changes for capitalization/case reasons. However, I did notice after that the strings are more or less consistent with what is currently in obs-deps, except maybe SpeexDSP, so I've dropped those from this review for now. I'm just making a note of it here as food for thought for a later date, since I don't want to further complicate this PR with _more_ string nitpicks than I have already included.

I _think_ I tried to keep the capitalisation in line with how either the projects call their libraries, but it's a bit hit and miss on non-case sensitive file systems.

> Most of the Windows native build scripts do not have their own Switch Parameters. I think it would be useful if they did so that if they are invoked directly, the Switch Parameters should be hinted on the CLI.

Dependency scripts are not meant to be invoked directly - if a single dependency is to be built it should be passed to the main build script via the `-Dependencies` switch (takes a comma separated string). The reason those are `param` blocks is actually for visual purposes only, but not a technical requirement.

Again, that's to keep code complexity down.

> This PR currently only builds ajantv2.lib in Release configuration. We currently build both Release and Debug builds of this library.

It's messy, but I guess adding an extra step to the Workflow to build the debug version before building the rest as release _should_ work.

PatTheMav commented on .Aptfile:
```diff
@@ -0,0 +1,11 @@
+package 'pkg-config'
+package 'wget'
+package 'yasm'
+package 'ccache'
```
It actually is by accident, only ccache needs to be moved up. 😂

PatTheMav commented on .github/actions/build-ffmpeg/action.yml:
```diff
@@ -0,0 +1,77 @@
+      run: |
+        case "${{ runner.os }}" in
+          Linux)
+            if ! type zsh > /dev/null 2>&1; then
```
`type` checks whether the binary exists on the system to avoid an unnecessary `apt` install run.

RytoEX replied:
Got it.

PatTheMav commented on .github/actions/build-ffmpeg/action.yml:
```diff
@@ -0,0 +1,77 @@
+name: 'Build FFmpeg'
+description: 'Builds FFmpeg for obs-deps with specified architecture, type, and build config'
```
IMO the usage is correct:

* The `description` field states what the action _achieves_ (it only exists in the file)
* The `name` field is printed to state what the workflow currently _does_

The description is never shown during a workflow run, the `name`s are however and serve as a "current status" display, hence the different verbiage.

RytoEX replied:
Works for me.

PatTheMav commented on .github/actions/build-macos-qt/action.yml:
```diff
@@ -0,0 +1,48 @@
+description: 'Builds macOS Qt for obs-deps with specified architecture and build config'
+inputs:
+  target:
+    description: 'Build target for dependencies'
```
Fixed.

PatTheMav commented on .github/actions/build-ffmpeg/action.yml:
```diff
@@ -0,0 +1,77 @@
+    required: false
+    default: 'Release'
+  cacheRevision:
+    description: 'Cache revision number to invalidate current cache generation'
```
Rephrased it to "Cache revision number to force creation of new cache generation".

PatTheMav commented on .github/actions/build-ffmpeg/action.yml:
```diff
@@ -0,0 +1,77 @@
+          !${{ github.workspace }}/*_build_temp/ffmpeg*/.git
+        key: ${{ inputs.target }}-ffmpeg-${{ inputs.type }}-${{ steps.ffmpeg-env-setup.outputs.hash }}-${{ inputs.cacheRevision }}
+
+    - name: Install FFmpeg
+      if: ${{ steps.ffmpeg-cache.outputs.cache-hit == 'true' }}
+      shell: zsh {0}
+      run: ./build-ffmpeg.zsh --skip-build --skip-unpack --target ${{ inputs.target }} --config ${{ inputs.config }} --${{ inputs.type }}
+
+    - name: Build FFmpeg
+      if: ${{ steps.ffmpeg-cache.outputs.cache-hit != 'true' }}
+      shell: zsh {0}
+      run: ./build-ffmpeg.zsh --target ${{ inputs.target }} --config ${{ inputs.config }} --${{ inputs.type }}
```
Implicitly yes - I guess it makes sense to explicitly call that out in the `name`.

PatTheMav commented on .github/actions/build-ffmpeg/action.yml:
```diff
@@ -0,0 +1,77 @@
+  type:
+    description: 'Build type (shared or static libraries)'
+    required: false
+    default: 'static'
```
I haven't heard any definite call on this - both work, so I'm happy to change it to whatever is required by the team.

PatTheMav replied:
Also forgot to add - that is an action parameter, so it can be set explicitly from the workflow.

RytoEX replied:
I'd ask @notr1ch what the preference is for static vs. shared builds of FFmpeg.

I did notice that it's an action parameter. I'd just like the default value to reflect the project's preferences.

PatTheMav replied:
Pinged R1ch on Discord.

RytoEX replied:
Sounds good. Resolve this conversation when you're satisfied then.

PatTheMav commented on .github/actions/build-windows-deps/action.yml:
```diff
@@ -0,0 +1,67 @@
+      id: deps-env-setup
+      shell: pwsh
+      run: |
+        Get-Content .\deps.windows\*.ps1 > temp.txt
+        $DepsHash = ((Get-FileHash -Path temp.txt -Algorithm SHA256).Hash)
```
Tbh I wasn't even aware of the alternative, I just wanted to get that workflow to work while coding it late at night. 😬

RytoEX replied:
If you consider this saner than the alternatives and "good enough", then feel free to resolve this conversation on your next pass.

PatTheMav replied:
Yeah I'll have to run some checks with it first - probably "good enough" for now.

PatTheMav commented on .github/actions/build-windows-deps/action.yml:
```diff
@@ -0,0 +1,67 @@
+
+        ./Build-Dependencies.ps1 @Params
+
+    - name: Build macOS Dependencies
```
Fixed.

PatTheMav commented on .github/actions/create-universal/action.yml:
```diff
@@ -0,0 +1,92 @@
+name: 'Create Universal Build'
```
Yeah makes sense, fixed.

PatTheMav commented on .github/workflows/main.yml:
```diff
@@ -13,660 +13,513 @@ on:
+            config: 'Release'
+            type: 'static'
+          - target: linux-x86_64
+            os: 'ubuntu-latest'
+            config: 'Release'
+            type: 'static'
+          - target: windows-x64
+            os: 'ubuntu-latest'
+            config: 'Release'
+            type: 'static'
+          - target: windows-x86
+            os: 'ubuntu-latest'
```
Good point, fixed it to use ubuntu-20.04 by default.

PatTheMav commented on .github/workflows/main.yml:
```diff
@@ -13,660 +13,513 @@ on:
+            ${{ matrix.target }}-ccache-ffmpeg-
 
-      - name: 'Setup build environment'
+      - name: Check for Github Labels
```
Fixed.

PatTheMav commented on .gitignore:
```diff
@@ -2,14 +2,14 @@
 ._*
 # Ignore Brew bundle files
-Brewfile.lock.json
+.Brewfile.lock.json
```
Brew bundler inherits the name of the Brewfile it is invoked with (which by default is `.Brewfile` in the home directory). To put more emphasis on the actual script files in the project root, I chose to name the meta files with a leading dot (`.Brewfile`, `.Aptfile`, `.Wingetfile`) as those are hidden by default.

RytoEX replied:
> I chose to name the meta files with a leading dot (.Brewfile, .Aptfile, .Wingetfile) as those are hidden by default.

I don't think this is true on Windows. The rest of the reasoning is good enough for me, I just wanted to confirm the filename pattern.

PatTheMav commented on Build-Dependencies.ps1:
```diff
@@ -0,0 +1,176 @@
+function Run-Stages {
+    $Stages = @('Setup')
+
+    if ( ( $script:SkipAll.isPresent ) -or ( $script:SkipBuild.isPresent ) ) {
```
Yeah I removed passing variables into imported scripts to reduce code complexity and explicitly use the `script` scope to access the value of switches passed to the main build scripts.

As such `$script:SkipAll` is correct, but per the docs `isPresent` could be removed.

PatTheMav replied:
Looks like removing `isPresent` breaks the script in some places, but not others. Looks like its return value is more compatible that whatever the variable itself coalesces to. I reverted the changes accordingly.

RytoEX replied:
Per the docs:
> Switch parameters are of type `SwitchParameter`, which implicitly converts to Boolean.

What's interesting is that `.IsPresent` is meant only for SwitchParameters, and a $script scope variable is not necessarily a Switch Parameter, so this should be an unreliable call as-is.

I think the more reliable method in cases like this would be to use the $script scope variable to set the default value of the Switch Parameter of this file like we do in the current build scripts:
```
[switch] $SkipAll = $(if (Test-Path Variable:SkipAll) { $SkipAll }),
```

PatTheMav replied:
Correct, the switch is defined in this file, so the `script` prefix is unnecessary here. In dot-included files the prefix has to be used as it needs to "inherit" the switch value from the calling script scope (and it seems that only the return value of `IsPresent` can be used as an array index).

PatTheMav commented on deps.ffmpeg/patches/FFmpeg/0003-FFmpeg-4.4.1-libaomenc.patch:
```diff
@@ -16,4 +16,3 @@ index e739cd3b66..b3d458a686 100644
          return AVERROR(EINVAL);
 -- 
 2.25.1
-
```
I think so? Patch is notoriously sensitive about trailing newlines.

RytoEX replied:
Apparently, that's a git format-patch signature. It could technically be omitted, but let's keep the patch changes minimal. If this whitespace change was needed, then fine.

PatTheMav commented on build-deps.zsh:
```diff
@@ -0,0 +1,136 @@
+setopt NO_AUTO_PUSHD
+setopt NO_PUSHD_IGNORE_DUPS
+setopt FUNCTION_ARGZERO
+#setopt WARN_CREATE_GLOBAL
+#setopt WARN_NESTED_VAR
+#setopt XTRACE
```
Those are default shell options helpful for debugging and development. I kept them in there so that maintainers/contributors can easily enable them to debug any issues that might occur.

To answer the last question, those should be familiar to anyone creating scripts in Zsh.

RytoEX replied:
Got it.

PatTheMav commented on deps.ffmpeg/10-zlib.zsh:
```diff
@@ -0,0 +1,97 @@
+name='zlib'
+version='1.2.11'
+url='https://github.com/madler/zlib.git'
+hash="cacf7f1d4e3d44d871b605da3b647f07d718623f"
```
Fixed.

PatTheMav commented on deps.ffmpeg/10-zlib.zsh:
```diff
@@ -0,0 +1,97 @@
+    cd ${target_config[output_dir]}
+
+    autoload -Uz apply_patch
+    apply_patch "${funcsourcetrace[1]:A:h}/patches/zlib/0001-disable-unistd-import.patch" \
```
The difference is mainly the line number in their output (absolute vs. relative). For our purpose it should be the same, but I'd need to test that first.

RytoEX replied:
A brief glance at the zsh documentation says, "Most users will be interested in the information in the funcfiletrace array instead." There is nothing said about if one is more performant, so I'll leave this call up to you.

PatTheMav replied:
I did run a test and - as the zsh docs vaguely mention - the file name references can be different. So changing it to `funcfiletrace` broke it. ¯\_(ツ)_/¯

PatTheMav commented on deps.ffmpeg/patches/librist/0001-generate-cross-compile-files-macos.patch:
```diff
@@ -0,0 +1,47 @@
++
++[built-in options]
++c_std = 'c11'
++cpp_std = 'c++11'
```
I think that's more or less a default target, I haven't checked if it supports C++17.

RytoEX replied:
Ah, librist seems to be C-only, and uses C99 as a default

rist-cpp, the C++ wrapper for librist, uses C++17, according to https://code.videolan.org/rist/rist-cpp:
> Requires cmake version >= 3.10, C++17, meson, ninja and googletest

However, the latter is not applicable here. Will let you decide what's best to specify here.

PatTheMav commented on deps.ffmpeg/patches/librist/0001-generate-cross-compile-files-windows.patch:
```diff
@@ -12,18 +12,24 @@ index 0000000..aa1e0fe
 +[binaries]
 +c = 'i686-w64-mingw32-gcc'
-+cpp = 'i686-w64-mingw32-c++'
++cpp = 'i686-w64-mingw32-g++'
```
I haven't had any issues with it, but I also think @pkviet generated the Windows cross compile files.

RytoEX replied:
Is this based on an older version of the initial librist PR then? I had him change from w64-mingw32-g++ to w64-mingw32-c++. All of the current deps build scripts should be the same way.

https://github.com/obsproject/obs-deps/pull/86#discussion_r776527267

PatTheMav commented on deps.ffmpeg/30-libogg.zsh:
```diff
@@ -0,0 +1,119 @@
+  windows 31bd3f2707fb7dbae539a7093ba1fc4b2b37d84e
+)
+local -a patches=(
+  "windows ${0:a:h}/patches/libogg/0001-fix-library-output-name.patch \
+    d03f003a186422247516022bd6d83fd7041c24bc8c0381ad9c43314b35c4c536"
```
I've added descriptions to a few patch files iirc, this one states https://github.com/msys2/MINGW-packages/issues/5825 as the origin of the patch.

There is no official commit or patch by Xiph themselves, mingw seems to be fixing this via patching as well.

RytoEX replied:
I wonder if that's resolved by any of the PRs/MRs on the ogg repos (GitHub and GitLab). Though, I have no idea how maintained ogg is these days.

RytoEX replied:
At least the patch has a description then. I just missed it the first time. Until Xiph updates ogg, this will have to suffice.

PatTheMav commented on deps.ffmpeg/30-libogg.zsh:
```diff
@@ -0,0 +1,119 @@
+autoload -Uz log_debug log_error log_info log_status log_output
+
+name='libogg'
+local -A versions=(
+  macos 1.3.5
+  linux 1.3.5
+  windows 1.3.4
+)
+local -A urls=(
+  macos https://github.com/xiph/ogg/releases/download/v1.3.5/libogg-1.3.5.tar.xz
+  linux https://github.com/xiph/ogg/releases/download/v1.3.5/libogg-1.3.5.tar.xz
+  windows https://github.com/xiph/ogg.git
+)
+local -A hashes=(
+  macos "${0:a:h}/checksums/libogg-1.3.5.tar.xz.sha256"
+  linux "${0:a:h}/checksums/libogg-1.3.5.tar.xz.sha256"
+  windows 31bd3f2707fb7dbae539a7093ba1fc4b2b37d84e
+)
+local -a patches=(
```
It's actually the other way around - most correctly use `local`, the ones that don't are the ones that need to be fixed.

It doesn't make too much of a difference, as variables are `local` scoped in Zsh by default, but I'll do a pass over the scripts to bring them all in line.

PatTheMav commented on deps.ffmpeg/40-aom.zsh:
```diff
@@ -0,0 +1,118 @@
+autoload -Uz log_debug log_error log_info log_status log_output
+
+name='aom'
+local version='3.2.0'
```
See above - it just makes the default Zsh behaviour more explicit. They are actually both `local`.

PatTheMav commented on deps.ffmpeg/60-lame.zsh:
```diff
@@ -0,0 +1,144 @@
+local -a patches=(
+  "* ${0:a:h}/patches/lame/0001-remove-outdated-symbol.patch \
+    d065b95e938652a6c219df7c9c057ba73e23e60fabb42cca633304ebed87a176"
+  "windows ${0:a:h}/patches/lame/0002-enable-ldflags-support-for-shared-libs.patch \
+    8beb0a98f15f8a0a935f9d68ad46aaa13cef9b24b2e34879201c81f39c72b5d8"
```
Yes, I just added `targets=('macos-*' 'linux-*')` to enable that behaviour.

PatTheMav commented on utils.zsh/functions/dep_checkout:
```diff
@@ -0,0 +1,71 @@
+} else {
+  log_debug 'No local checkout found'
+  local -a git_version=($(git --version 2>/dev/null))
+  if (( sparse )) && is-at-least 2.25 ${git_version[3]}; then
+    log_debug "Attempt sparse checkout with options ${args}"
+    git clone --filter=blob:none --no-checkout "${url}" "${PWD}"
+    git sparse-checkout ${args}
+  else
+    log_debug "Clone repository from ${url}"
+    git clone "${url}" "${PWD}"
+  fi
```
The short form curly brackets only works if the condition is enclosed in `(( ))` or `[[ ]]`. That's a Zsh convention.

RytoEX replied:
Would it make sense to make them consistent in all usages, or is that generally considered unnecessary in zsh scripts?

PatTheMav commented on deps.ffmpeg/99-ffmpeg.zsh:
```diff
@@ -0,0 +1,294 @@
+    --enable-libsrt
+    --enable-shared
+    --disable-static
+    --disable-libjack
+    --disable-indev=jack
+    --disable-outdev=sdl
```
SDL maybe, but this is more important for POSIX systems where both dependencies might otherwise exist on the system and will be picked up by FFmpeg. Doesn't hurt to have those by default.

PatTheMav commented on deps.macos/qt.zsh:
```diff
@@ -0,0 +1,343 @@
+)
+
+setup() {
+  if [[ ${shared_libs} -eq 0 && ${CPUTYPE} != "${arch}" ]] \
+    { log_error "Cross compilation requires shared library build"; exit 2 }
```
Yeah it's a toss-up - I made it more readable.

RytoEX replied:
The style guide we follow for OBS doesn't permit multiple statements on one line, so I usually defer to that as a style preference across the project.

PatTheMav commented on deps.windows/10-mbedtls.ps1.disabled:
```diff
@@ -0,0 +1,90 @@
+    $OnOff = @('OFF', 'ON')
+    $Options = @(
+        $CmakeOptions
+        "-DUSE_SHARED_MBEDTLS_LIBRARY=$($OnOff[$Shared.isPresent])"
```
Yeah see above, I'll do a pass over the Powershell scripts and remove the `isPresent` calls.

RytoEX replied:
Since this is actually a Switch Parameter for this script, removing `.IsPresent` here should work. However, I do notice that this is different from the other scripts where you do not use a Switch Parameter, and instead use a $script scope variable. So perhaps you meant this to be $script scope and not have a Switch Parameter present for this script file?

PatTheMav commented on deps.windows/10-mbedtls.ps1.disabled:
```diff
@@ -0,0 +1,90 @@
+
+function Patch {
+    Log-Information "Patch (${Target})"
+
```
Fine by me.

PatTheMav commented on deps.windows/20-freetype.ps1:
```diff
@@ -0,0 +1,94 @@
+    $Options = @(
+        $CmakeOptions
+        "-DBUILD_SHARED_LIBS=$($OnOff[$script:Shared.isPresent])"
+        '-DFT_WITH_BROTLI=OFF'
+        '-DFT_WITH_BZIP2=OFF'
+        '-DFT_WITH_HARFBUZZ=OFF'
+        '-DFT_WITH_PNG=OFF'
+        '-DFT_WITH_ZLIB=OFF'
```
I actually tried to disable the spam generated by CMake because it cannot find any of these on Windows - it seems it still tries to find the libraries (even if I explicitly disable them), but I also tried to be explicit about disabling anything we don't need.

RytoEX replied:
Fair enough. Maybe FreeType's CMake for these needs some work? Something to look into another time.

PatTheMav commented on deps.windows/20-freetype.ps1:
```diff
@@ -0,0 +1,94 @@
+        Move-Item @Item
+    }
+
+    Remove-Item -ErrorAction 'SilentlyContinue' -Recurse "$($ConfigData.OutputPath)/lib/cmake/freetype"
```
In general build scripts are supposed to "clean up after themselves", because that knowledge is closely tied to the dependency itself (it also makes the packaging step simpler as you don't have to add new cleanup steps there).

For this PR I tried to mimic the resulting windows-deps package as much as possible. It would _actually_ make sense to keep the cmake directories, but we don't right now.

RytoEX replied:
For the current build scripts, I had elected to leave cleanup to the packaging step to leave the intermediate artifacts from cross-compiling as unmodified as possible. I'll leave this up to you.

PatTheMav replied:
Due to the import library fix from yesterday I actually opted _not_ to remove cmake and pkg-config files anymore, which made the entire fixup step unnecessary for quite a few scripts anyway.

PatTheMav commented on deps.ffmpeg/99-ffmpeg.zsh:
```diff
@@ -0,0 +1,294 @@
+## Dependency Information
+name='ffmpeg'
+local version='4.4.1'
+local url='https://github.com/ffmpeg/ffmpeg.git'
```
The name is used for display purposes, so "FFmpeg" makes sense there, but file paths and URLs are case-insensitive anyway.

RytoEX replied:
URL paths are case-sensitive. GitHub just makes it appear as though they are not. The Canonical URL per the GitHub repo's HTML source is "https://github.com/FFmpeg/FFmpeg", and when getting the git repo URL from GitHub, the case is as I specified.

PatTheMav commented on deps.ffmpeg/30-libvpx.zsh:
```diff
@@ -0,0 +1,186 @@
+autoload -Uz log_debug log_error log_info log_status log_output
+
+## Dependency Information
```
Sounds good, added those.

PatTheMav commented on deps.windows/10-mbedtls.ps1.disabled:
```diff
@@ -0,0 +1,90 @@
+
+function Install {
+    Log-Information "Install (${Target})"
+
```
Yeah streamlined the formatting.

PatTheMav commented on deps.windows/20-freetype.ps1:
```diff
@@ -0,0 +1,94 @@
+
+function Fixup {
+    Log-Information "Fixup (${Target})"
+
```
See above.

PatTheMav commented on utils.pwsh/Bootstrap.ps1:
```diff
@@ -0,0 +1,38 @@
+        Write-Host -ForegroundColor Green 'Yes'
+    } else {
+        Write-Host -ForegroundColor Red 'No'
+
```
Fixed.

PatTheMav commented on utils.pwsh/Bootstrap.ps1:
```diff
@@ -0,0 +1,38 @@
+}
+
+$Self = $MyInvocation.MyCommand.Name
+
```
Fixed.

PatTheMav commented on utils.pwsh/Bootstrap.ps1:
```diff
@@ -0,0 +1,38 @@
+function Bootstrap {
+    <#
+        .SYNOPSIS
+            Bootstraps an OBS-deps build environment
```
Fixed.

PatTheMav commented on utils.pwsh/Check-Git.ps1:
```diff
@@ -0,0 +1,25 @@
+        .SYNOPSIS
+            Ensures available git executable on host system
+        .DESCRIPTION
+            Checks whether a git command is available on the host system. If non is found,
```
Fixed.

PatTheMav commented on utils.pwsh/Ensure-Location.ps1:
```diff
@@ -0,0 +1,29 @@
+        .SYNOPSIS
+            Ensure current location to be set to specified directory
+        .DESCRIPTION
+            If specified directory exists, switch to it. Otherweise create it,
```
Fixed.

PatTheMav commented on utils.pwsh/Bootstrap.ps1:
```diff
@@ -0,0 +1,38 @@
+function Bootstrap {
+    <#
+        .SYNOPSIS
+            Bootstraps an OBS-deps build environment
```
Fixed.

PatTheMav commented on utils.pwsh/Expand-ArchiveExt.ps1:
```diff
@@ -0,0 +1,55 @@
+        .SYNOPSIS
+            Expands archives of wide support of file formats
+        .DESCRIPTION
+            Allows extraction of zip archives, as well as 7z, gz and xz archives.
```
Oxford comma strikes again - fixed.

PatTheMav commented on utils.pwsh/Expand-ArchiveExt.ps1:
```diff
@@ -0,0 +1,55 @@
+            }
+            break
+        }
+        .gz {
+            Invoke-External tar -x -o $DestinationPath -f $Path
+            break
+        }
+        .xz {
+            Invoke-External tar -x -o $DestinationPath -f $Path
+            break
```
Good point - added the fallbacks.

PatTheMav commented on utils.pwsh/Install-BuildDependencies.ps1:
```diff
@@ -0,0 +1,59 @@
+            Install required build dependencies
+        .DESCRIPTION
+            Additional packages might be needed for successful builds. This module contains additional
+            dependencies available for installation via winget and - if possible - adds their locations
```
Fixed.

PatTheMav commented on utils.pwsh/Invoke-DevShell.ps1:
```diff
@@ -0,0 +1,79 @@
+        .SYNOPSIS
+            Invoke a VSDevShell as a subshell and run build commands in it
+        .DESCRIPTION
+            To avoid polluting the host Powershell environment, this function allows the dynamic
+            creation of a scriptblock that runs inside a Powershell subshell, which will find
+            and setup a Visual Studio Dev Shell for the selected target architecture and start
+            the supplied build command.
```
Fixed.

PatTheMav commented on utils.pwsh/Invoke-DevShell.ps1:
```diff
@@ -0,0 +1,79 @@
+        $PowerShellCommand = "powershell"
+    }
+
+    #$DevScriptBlock = [ScriptBlock]::Create($DevShellCommand)
```
I don't think so, it might be a remnant when I struggled with compositing the command string. I removed it for now.

PatTheMav commented on utils.pwsh/Invoke-External.ps1:
```diff
@@ -0,0 +1,40 @@
+function Invoke-External {
+    <#
+        .SYNOPSIS
+            Invoke a non-Powershell command
```
(╯'□')╯︵ ┻━┻ 

Fixed.

PatTheMav commented on utils.pwsh/Invoke-SafeWebRequest.ps1:
```diff
@@ -0,0 +1,79 @@
+            a corresponding FileHash object file is used to confirm the
+            contents of the downloaded file.
+        .EXAMPLE
+            Invoke-SafeWebRequest -Uri "My-Url" -HashFile "Path-To-HashFile"
+            Invoke-SafeWebRequest -Uri "My-Url" -HashFile "Path-To-HashFile" -Resume
+            Invoke-SafeWebRequest -Uri "My-Url" -HashFile "Path-To-HashFile" -OutFile "Name-Assumed-By-HashFile"
```
Yep PowerShell uses `Uri` throughout, that should match.

PatTheMav commented on utils.pwsh/Setup-Target.ps1:
```diff
@@ -0,0 +1,95 @@
+
+    Log-Debug @"
+
+Cmake options   : $($script:CmakeOptions)
```
Fixed.

PatTheMav commented on utils.pwsh/Setup-Target.ps1:
```diff
@@ -0,0 +1,95 @@
+            Find-VisualStudio
+    #>
+
+    $VisualStudioData = Get-CimInstance MSFT_VSInstance
```
Microsoft's Docs are unusually coy about that, I found suggestions that it can return _all_ available instances, but I haven't had time to create such an environment (especially as it isn't expected on CI). Given that the call returns a PowerShell object it should be easy to filter/adapt if that situation occurs.

RytoEX replied:
I guess we'll just leave this be until someone runs into it.

PatTheMav commented on utils.pwsh/Utils-Patch.ps1:
```diff
@@ -0,0 +1,142 @@
+
+    $PatchExe = "${GitBasePath}\patch.exe"
+
+    # Invoke-External $PatchExe --binary @args
```
I kept it as it should be reinstated once the PowerShell version that doesn't require the entire workaround for running `patch.exe` is made available on CI and officially released.

PatTheMav replied:
Here we go - issue is supposedly fixed in 7.3.0-preview2, released 17 days ago: https://github.com/PowerShell/PowerShell/pull/16612#issuecomment-1050363883

RytoEX replied:
Would it make sense to provide some context that the commented out command would work in 7.3.0+ or use a version check? Leaving it commented with no context is bound to end up with it forgotten.

PatTheMav replied:
I opt for the version check, I'll implement that later today.

RytoEX replied:
Looks like you pushed this change, so I'll mark this resolved.

PatTheMav commented on deps.ffmpeg/60-srt.zsh:
```diff
@@ -0,0 +1,142 @@
+  args=(
+    ${cmake_flags}
+    -DENABLE_SHARED="${_onoff[(( shared_libs + 1 ))]}"
+    -DENABLE_STATIC=ON
```
Yeah pkv asked me the same - both can be enabled at the same time. We always want static builds and optionally also get shared on top.

For some projects this also influences which libraries get installed by `cmake --install`.

PatTheMav commented on utils.zsh/10-bootstrap.zsh:
```diff
@@ -0,0 +1,168 @@
+autoload -Uz is-at-least && if ! is-at-least 5.2; then
+  print -u2 -PR "%F{1}${funcstack[1]##*/}:%f Running on Zsh version %B${ZSH_VERSION}%b, but Zsh %B5.2%b is the minimum supported version. Upgade Zsh to fix this issue."
```
Fixed.

PatTheMav commented on deps.ffmpeg/60-mbedtls.zsh:
```diff
@@ -0,0 +1,172 @@
+
+  args=(
+    ${cmake_flags}
+    -DUSE_SHARED_MBEDTLS_LIBRARY="${_onoff[(( shared_libs + 1 ))]}"
+    -DUSE_STATIC_MBEDTLS_LIBRARY=ON
```
See above, we always want static builds (so we specify it by default) and disable shared builds based on preference. These values aren't implicitly tied with each other (otherwise it would be a single static/shared switch).

RytoEX replied:
Got it. I'll have to check how this works in practice with building the final Windows deps package.

PatTheMav commented on utils.zsh/20-host.zsh:
```diff
@@ -0,0 +1,113 @@
+  log_info "Checking for ccache..."
+
+  if (( ${+commands[ccache]} )) {
+    log_status "Ccache found"
```
I'm actually unsure what the correct capitalisation of "ccache" is.

RytoEX replied:
Checked, and this usage is fine.

PatTheMav commented on utils.zsh/30-build.zsh:
```diff
@@ -0,0 +1,230 @@
+        "${target_config[cross_prefix]}-w64-mingw32-c++"
+        "${target_config[cross_prefix]}-w64-mingw32-cpp"
+        "${target_config[cross_prefix]}-w64-mingw32-gcc"
+        "${target_config[cross_prefix]}-w64-mingw32-g++"
```
That's for completeness' sake - the script checks for all available compiler aliases made available via ccache.

RytoEX replied:
I guess my question was more, if we have been specifically using mingw32-c++ and avoiding mingw32-g++, should we list it here? I'm not sure what this is doing or used for.

PatTheMav replied:
It just prints the detected compiler paths for use with ccache if `--debug` is used.

RytoEX replied:
Resolved this in an off-thread conversation. I don't think we're actually using g++ anywhere, but this shouldn't hurt to leave in place.

PatTheMav commented on utils.zsh/40-dependency.zsh:
```diff
@@ -0,0 +1,50 @@
+
+%BOptions%b:
+
+  %Burl%b            Download url or git repository url
```
Fixed.

PatTheMav commented on utils.zsh/40-dependency.zsh:
```diff
@@ -0,0 +1,50 @@
+  %Bcommit%b         Git commit id for checkout (%F(yellow)mandatory for git checkouts%f)
+
+  %Bbranch%b         Branch to use for git checkout (%F(blue)optional%f)
+  %Bpull_request%b   Github-based pull request ID to use for git checkout (%F(blue)optional%f)"
```
Fixed.

PatTheMav commented on utils.zsh/functions/check_cross:
```diff
@@ -0,0 +1,23 @@
+  case ${host_os} {
+    linux)
+      if (( ! ${+CI} )) log_warning 'Dependency installation via apt may require elevated privileges'
+      sudo apt -y install mingw-w64 mingw-w64-tools dpkg-dev ${_quiet[(( (_loglevel == 0) + 1 ))]}
```
`apt` is the recommended way - `apt-get` is a low-level command that shouldn't be used by users.

RytoEX replied:
My understanding is that `apt-get` is recommended for scripts, while `apt` is recommended for interactive use by human users. `apt` is also not guaranteed to have a stable CLI API or backwards compatibility. If you're not piping `apt` into other commands, it's probably safe enough.

PatTheMav replied:
Makes sense, updated it in this script and the others using `apt`.

pkviet commented:
Tested rist and srt dll with the native rist & srt PR. They work fine.

PatTheMav commented:
Added commits for the changes/fixes requested by @RytoEX - will merge them back later if we finished reviews.

RytoEX commented on deps.ffmpeg/99-ffmpeg.zsh:
```diff
@@ -0,0 +1,294 @@
+    --enable-libvorbis
+    --enable-libvpx
+    --enable-libtheora
+    --enable-libmp3lame
```
These are definitely not enabled in the currently used build scripts for Windows.

https://github.com/obsproject/obs-deps/blob/b5a42eff643e0474271d6a62aee708818af98239/CI/windows/build_ffmpeg.sh#L60-L85

It does look like you removed these from the windows targets though.

RytoEX commented:
> > For libvpx and libsrt, can we implement something to use custom filenames so that both the downloaded files/directories and the checksum files have more descriptive names?
> 
> Would add unnecessary code complexity, I deliberately did not implement code for that as it makes file name composition (download artifact, extraction folder, checksum file) quite a bit easier.

That's unfortunate, as it's not very clear what those files with just version numbers as names are for.

> > I noticed that `CI/patches/QTBUG-74606.patch` was moved/renamed to `deps.macos/patches/Qt/0001-QTBUG-74606.patch`. We use this patch for Windows too, even though we don't currently build Windows Qt on CI (I plan to address that after this lands). Would the expectation be to put this in a Windows-specific `deps.windows/patches` folder? Or do we have a platform-agnostic patch directory available with this new system?
> 
> That depends on how we build Qt for Windows. Ideally we'd have a platform-agnostic `deps.qt` folder that contains the patches. We can probably throw both Powershell as well as Zsh files in there without issue, so that's where all the Qt-related code would migrate to then.

I plan to work on the Qt build after this lands. The "how" right now for Qt5 is basically:
1. check out git repo
2. run their perl script `init-repository` with specific modules selected
3. configure
4. jom
5. jom install

Alternatively, one could replace 1 and 2 by downloading the _full_ source package, though I'd expect it to be considerably larger. Of special note is that for Windows, we build with `-debug-and-release` which seems to be the default on macOS and Windows in Qt 5.15.2:
> -debug-and-release ... Build two versions of Qt, with and without debugging turned on [yes] (Apple and Windows only)

The build instructions for Qt6 don't use jom, and just use `cmake --build` and `cmake --install` and invoke MSBuild/MSVC that way. They do still use `configure` though.

> > If possible, I would prefer that we keep all dependency versions the same as they are currently for this PR, unless there are problems with those dependencies with this build system. That way, we can cleanly separate and more easily track these build system changes from the dependency updates and make them easier to find in commit history. I've noted a few places in specific files where I think dependency updates were made.
> 
> That shouldn't be an issue architecture-wise, I just prefer _in general_ to keep our macOS deps on the most recent version supported by Homebrew, as that community does the heavy lifting of testing and fixing build issues for us. That approach (at least so far) has resulted in the least amount of headaches while at the same time allowing very regular updates on macOS compared to Windows.

If it is prohibitively difficult to keep the versions the same for this entire PR, then I'd be fine with separate commits for updating the versions (and associated hash files) after all of the build system changes. I'm not objecting to updating the versions entirely, I just don't want version update changes happening in build system change commits because it makes it more difficult to track what versions changed when.

> > The mbedTLS Windows native build is disabled. Is that related to the change to build FFmpeg components as static?
> 
> That's related to the mbedTLS version build as part of the FFmpeg build is shared by default and didn't have any issues in OBS on Windows in my testing (including RTMPs streaming to Youtube), so no need to build it twice.

I'll have to double check this later, as I recall the mbedTLS stuff being finicky, which is why the build process for it had multiple steps.

> > What is the minimum PowerShell version supported? As the GitHub Actions images for both Windows 2019 and Windows 2022 have PowerShell 7.2, and that is relatively easy to install on Windows 10, I'd be fine setting that as a required version if that makes things simpler.
> 
> I'd guess the earliest PowerShell Core version (pwsh), I'd guess going with PowerShell 7.2 would be easiest straight out the door.

Windows 10 comes installed with 5.1 by default. However, it tells you on load to:
> Try the new cross-platform PowerShell https://aka.ms/pscore6

Which leads to a page to install the latest stable PowerShell. Whatever we choose, the script should warn or error out if the minimum isn't met. I don't know what Windows 11 comes installed with.

> > Most of the Windows native build scripts do not have their own Switch Parameters. I think it would be useful if they did so that if they are invoked directly, the Switch Parameters should be hinted on the CLI.
> 
> Dependency scripts are not meant to be invoked directly - if a single dependency is to be built it should be passed to the main build script via the `-Dependencies` switch (takes a comma separated string). The reason those are `param` blocks is actually for visual purposes only, but not a technical requirement.
> 
> Again, that's to keep code complexity down.

Fair enough. Is that documented somewhere? That's a departure from the current system. I don't mind it, it was just unclear that that was the new expectation.

PatTheMav commented:
> That's unfortunate, as it's not very clear what those files with just version numbers as names are for.

It gets even worse - you have one name for the downloaded file, but because of the way the archives have been created you get a folder with the name of the archive and another properly named directory on top of that and solving that was already a headache. I might revisit this later, though switching to Git checkouts resolves this entirely.

> I plan to work on the Qt build after this lands. The "how" right now for Qt5 is basically:
> 
> 1. check out git repo
> 2. run their perl script `init-repository` with specific modules selected
> 3. configure
> 4. jom
> 5. jom install
> 
> Alternatively, one could replace 1 and 2 by downloading the _full_ source package, though I'd expect it to be considerably larger. Of special note is that for Windows, we build with `-debug-and-release` which seems to be the default on macOS and Windows in Qt 5.15.2:
> 
> > -debug-and-release ... Build two versions of Qt, with and without debugging turned on [yes] (Apple and Windows only)
> 
> The build instructions for Qt6 don't use jom, and just use `cmake --build` and `cmake --install` and invoke MSBuild/MSVC that way. They do still use `configure` though.

Sounds like it should be "easy" enough to migrate everything into a separate `deps.qt` folder then (patches included).

> If it is prohibitively difficult to keep the versions the same for this entire PR, then I'd be fine with separate commits for updating the versions (and associated hash files) after all of the build system changes. I'm not objecting to updating the versions entirely, I just don't want version update changes happening in build system change commits because it makes it more difficult to track what versions changed when.

In general I don't see an issue with synchronizing the versions in a later step and keeping them there, knowing that the framework supports different versions if necessary.

> I'll have to double check this later, as I recall the mbedTLS stuff being finicky, which is why the build process for it had multiple steps.

Yeah if you could dog-food those use cases that'd be great.

> Windows 10 comes installed with 5.1 by default. However, it tells you on load to:
> 
> > Try the new cross-platform PowerShell https://aka.ms/pscore6
> 
> Which leads to a page to install the latest stable PowerShell. Whatever we choose, the script should warn or error out if the minimum isn't met. I don't know what Windows 11 comes installed with.
 

Then let's go with Powershell Core v7, I'll check how to add the version check.

> Fair enough. Is that documented somewhere? That's a departure from the current system. I don't mind it, it was just unclear that that was the new expectation.

When you run the script with `-?` it will show you the available parameters:

```
Build-Dependencies.ps1 [[-Target] <string>] [[-Configuration] <string>] [[-Dependencies] <string[]>] [-SkipAll] [-SkipDeps] [-SkipUnpack] [-SkipBuild] [-Shared] [-Clean] [-Quiet] [<CommonParameters>]
```

If you "speak" pwsh, this tells you that `Dependencies` takes an array of strings, `Target` and `Configuration` is a constrained set, so you can tab-complete through the available values.

PatTheMav commented on build-deps.zsh:
```diff
@@ -0,0 +1,136 @@
+    stages+=(install fixup)
+  } else {
+    stages=(setup)
+    if (( clean_build )) stages+=(clean)
```
I guess it's a matter of taste - the zsh framework I help out with from time to time mandates this in its style guides: https://github.com/zimfw/zimfw/wiki/Code-Style-Guide

> For short single commands, prefer one-liners

RytoEX commented on utils.pwsh/Setup-Target.ps1:
```diff
@@ -0,0 +1,95 @@
+function Find-VisualStudio {
+    <#
+        .SYNOPSIS
+            Find available Visual Studio instance
```
Find -> Finds

RytoEX commented on utils.pwsh/Setup-Target.ps1:
```diff
@@ -0,0 +1,95 @@
+            Find available Visual Studio instance
+        .DESCRIPTION
+            Uses WMI (Windows Management Instrumentation) to find an installed
+            Visual Studio instance on the host system
```
Use punctuation.

RytoEX commented on utils.pwsh/Setup-Target.ps1:
```diff
@@ -0,0 +1,95 @@
+        Arch = ('x86', 'x64')[$Target64Bit]
+        UnixArch = ('x86', 'x86_64')[$Target64Bit]
+        CmakeArch = ('Win32', 'x64')[$Target64Bit]
+        Bits = ('32', '64')[$Target64Bit]
```
I prefer "Bitness".

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-CH-UA-Bitness
https://en.wiktionary.org/wiki/bitness
https://docs.rs/os_info/3.2.0/os_info/enum.Bitness.html
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-coma/3f01a312-3391-4109-a96e-95358babf191

RytoEX commented on utils.pwsh/Setup-Target.ps1:
```diff
@@ -0,0 +1,95 @@
+            "A Visual Studio installation (2019 or newer) is required for this build script.",
+            "The Visual Studio Community edition is available for free at https://visualstudio.microsoft.com/vs/community/.",
+            "",
+            "(If Visual Studio is indeed installed, locate the directory ",
+            " 'C:\ProgramData\Microsoft\VisualStudio\Packages\Microsoft.VisualStudio.Setup.WMIProvider,Version=xxxx'",
+            " right-click the file 'Microsoft.Visualstudio.Setup.WMIProvider.msi' and choose 'repair'.)"
```
Omit parentheses for this sentence.

RytoEX commented on utils.pwsh/Utils-Patch.ps1:
```diff
@@ -0,0 +1,145 @@
+        .SYNOPSIS
+            Applies a patch
+        .DESCRIPTION
+            Allows applying the changes contained in a patch file
```
Use punctuation.

RytoEX commented on utils.pwsh/Utils-Patch.ps1:
```diff
@@ -0,0 +1,145 @@
+function Revert-Patch {
+    <#
+        .SYNOPSIS
+            Reverts a patch
```
Use punctuation.

RytoEX commented on utils.pwsh/Utils-Patch.ps1:
```diff
@@ -0,0 +1,145 @@
+function Apply-Patch {
+    <#
+        .SYNOPSIS
+            Applies a patch
```
Use punctuation.

RytoEX commented on utils.pwsh/Utils-Patch.ps1:
```diff
@@ -0,0 +1,145 @@
+function Safe-Patch {
+    <#
+        .SYNOPSIS
+            Applies a patch with hash checking
```
Use punctuation.

RytoEX commented on utils.pwsh/Utils-Patch.ps1:
```diff
@@ -0,0 +1,145 @@
+    } else {
+        Pop-Location -Stack SafePatchTemp
+
+        throw "Hash of patch file ${PatchFile} is '${PatchHash}' - expected '${HashSum}'"
```
I'd kind of prefer two sentences (". Expected "), or ", but expected" rather than a hyphen.

RytoEX commented on utils.zsh/10-bootstrap.zsh:
```diff
@@ -0,0 +1,168 @@
+        ;;
+      -t|--target)
+        if (( ! ${_valid_targets[(Ie)${2}]} )) {
+          log_error "Invaid value %B${2}%b for option %B${1}%b"
```
Invaid -> Invalid

Fix here and throughout.

RytoEX commented on utils.zsh/functions/check_cross:
```diff
@@ -0,0 +1,23 @@
+
+log_debug 'Checking for apt-get...'
+if (( ! ${+commands[apt-get]} )) {
+  log_error 'No apt command found - please install apt'
```
'No apt-get command found. Please install apt.'

While apt-get is the command, apt is the package.

Though, I have no idea how that would occur.

PatTheMav replied:
It should not, but the check is more "complete" that way.

RytoEX replied:
Sure. I don't mind keeping the check. Resolve once the string change is addressed.

RytoEX commented on utils.zsh/functions/check_cross:
```diff
@@ -0,0 +1,23 @@
+  log_error 'No apt command found - please install apt'
+  return 2
+} else {
+  log_debug "Apt located at ${commands[apt-get]}"
```
Apt -> Apt-get

This casing is a bit weird, but avoiding it requires more words or just not capitalizing it.

RytoEX commented on utils.zsh/functions/check_deps:
```diff
@@ -0,0 +1,42 @@
+autoload -Uz log_info log_status log_error log_debug log_warning
+
+log_debug 'Checking for apt...'
```
apt -> apt-get

RytoEX commented on utils.zsh/functions/check_deps:
```diff
@@ -0,0 +1,42 @@
+
+log_debug 'Checking for apt...'
+if (( ! ${+commands[apt-get]} )) {
+  log_error 'No apt-get command found - please install apt'
```
'No apt-get command found. Please install apt.'

RytoEX commented on utils.zsh/functions/check_deps:
```diff
@@ -0,0 +1,42 @@
+  log_error 'No apt-get command found - please install apt'
+  return 2
+} else {
+  log_debug "Apt located at ${commands[apt-get]}"
```
Apt -> Apt-get

See check_cross for related notes.

RytoEX reviewed (changes requested):
Did another pass over the latest changes.

RytoEX commented:
> > That's unfortunate, as it's not very clear what those files with just version numbers as names are for.
> 
> It gets even worse - you have one name for the downloaded file, but because of the way the archives have been created you get a folder with the name of the archive and another properly named directory on top of that and solving that was already a headache. I might revisit this later, though switching to Git checkouts resolves this entirely.

That is frustrating, yeah. What's the tradeoff with Git checkouts versus file downloads? Data transfer, storage space?

> > I plan to work on the Qt build after this lands. The "how" right now for Qt5 is basically:
> > 
> > \<snip>
> 
> Sounds like it should be "easy" enough to migrate everything into a separate `deps.qt` folder then (patches included).

Would it be easier for us to coordinate on this later after this PR instead of addressing now?

> > If it is prohibitively difficult to keep the versions the same for this entire PR, then I'd be fine with separate commits for updating the versions (and associated hash files) after all of the build system changes. I'm not objecting to updating the versions entirely, I just don't want version update changes happening in build system change commits because it makes it more difficult to track what versions changed when.
> 
> In general I don't see an issue with synchronizing the versions in a later step and keeping them there, knowing that the framework supports different versions if necessary.

Let me know how this goes or if it's not possible.

> > I'll have to double check this later, as I recall the mbedTLS stuff being finicky, which is why the build process for it had multiple steps.
> 
> Yeah if you could dog-food those use cases that'd be great.

Will get back to you on this.

> > Windows 10 comes installed with 5.1 by default. However, it tells you on load to:
> > > Try the new cross-platform PowerShell https://aka.ms/pscore6
> > 
> > 
> > Which leads to a page to install the latest stable PowerShell. Whatever we choose, the script should warn or error out if the minimum isn't met. I don't know what Windows 11 comes installed with.
> 
> Then let's go with Powershell Core v7, I'll check how to add the version check.

Sounds good to me.

> > Fair enough. Is that documented somewhere? That's a departure from the current system. I don't mind it, it was just unclear that that was the new expectation.
> 
> When you run the script with `-?` it will show you the available parameters:
> 
> ```
> Build-Dependencies.ps1 [[-Target] <string>] [[-Configuration] <string>] [[-Dependencies] <string[]>] [-SkipAll] [-SkipDeps] [-SkipUnpack] [-SkipBuild] [-Shared] [-Clean] [-Quiet] [<CommonParameters>]
> ```
> 
> If you "speak" pwsh, this tells you that `Dependencies` takes an array of strings, `Target` and `Configuration` is a constrained set, so you can tab-complete through the available values.

Ah, I more meant if it was mentioned in the PR description and/or commit message. Something like, "Invocation of individual component scripts directly is no longer supported. Individual components can be built by invoking the main build script and specifying the component (examples for macOS, Windows native, and Windows cross)."

PatTheMav commented on .github/workflows/main.yml:
```diff
@@ -13,660 +13,526 @@ on:
+            macOS)
+              to_remove=()
+
+              for formula in llvm gcc postgresql openjdk sox libsndfile flac libvorbis opusfile \
+                libogg composer php gd freetype fontconfig webp libpng lame libtiff opus kotlin \
+                sbt libxft libxcb; do
```
That step uses bash, the second variant only works in Zsh.

RytoEX replied:
Ah, got it.

PatTheMav commented on .Aptfile:
```diff
@@ -0,0 +1,11 @@
+package 'ninja-build', bin: 'ninja'
+package 'pkg-config'
+package 'wget'
+package 'yasm'
```
Looks like a remnant of the macOS Brewfile.

PatTheMav commented on deps.ffmpeg/30-libvpx.zsh:
```diff
@@ -0,0 +1,188 @@
+## Dependency Information
+local name='libvpx'
+local -A versions=(
+  macos 1.11.0
+  linux 1.11.0
```
Yeah I think while updating x264 to coincide with the PR from a while ago I also updated some other deps on macOS - I'm fine with those and others if they are "tested" to compile without issue.

RytoEX replied:
Do you mean "compile without issue with these build scripts"? Because we compile 1.10.0 right now with the current build scripts.

PatTheMav replied:
Huh, so I can reply to your comment up here, but not further below. Curious.

Anyway, yes it's meant to say that within the scope of the PR it hasn't exhibited errors with 1.11.0 so far.

RytoEX replied:
It looks like this has been addressed.

PatTheMav commented on deps.ffmpeg/60-srt.zsh:
```diff
@@ -0,0 +1,133 @@
+
+## Dependency Information
+local name='srt'
+local version='1.4.4'
```
Yes, @pkviet asked to update to 1.4.4 and I obliged.

RytoEX replied:
As with the other updates, if that could be pushed into its own dependency update commit (individual per component, if possible; together as one dependency update commit if not), that would be great.

PatTheMav replied:
Good point, I'll extract those changes into separate commits then.

PatTheMav commented on deps.ffmpeg/20-opus.zsh:
```diff
@@ -0,0 +1,110 @@
+
+## Dependency Information
+local name='opus'
+local version='1.3.1'
+local url='https://github.com/xiph/opus.git'
+local hash="dfd6c88aaa54a03a61434c413e30c217eb98f1d5"
```
That's the macOS hash - the other one is almost 3 years old already.

PatTheMav commented on deps.windows/20-freetype.ps1:
```diff
@@ -0,0 +1,68 @@
+    [string] $Hash = '6a2b3e4007e794bfc6c91030d0ed987f925164a8'
+)
+
+function Setup {
+    Setup-Dependency -Uri $Uri -Hash $Hash -DestinationPath $Path
+}
```
I noticed that as well once I had all scripts done, for now I'd rather simplify this in a separate PR.

RytoEX replied:
Sure. I'm fine with that optimization happening at a later time.

PatTheMav commented on utils.pwsh/Invoke-SafeWebRequest.ps1:
```diff
@@ -0,0 +1,79 @@
+        . $PSScriptRoot/Utils-Logger.ps1
+    }
+
+    if ( $Resume -and $PSVersionTable.PSVersion -lt "6.1.0") {
+        Log-Warning "-Resume only available on Powershell 6.1.0 or later, disabling"
+        $Resume = $false
+    }
```
Not in the scope of the script, but the functions are written as generic drop-in scripts (e.g. you can source them for use in an interactive PowerShell session) so I felt it better to guard that feature explicitly behind that version number.

RytoEX replied:
Works for me.

RytoEX commented on Build-Dependencies.ps1:
```diff
@@ -0,0 +1,178 @@
+}
+
+if ( $PSVersionTable.PSVersion -lt '7.0.0' ) {
+    Write-Warning 'The obs-deps PowerShell build script requires PowerShell Core 7 - install or upgrade your PowerShell version: https://aka.ms/pscore6'
```
> The obs-deps PowerShell build script requires PowerShell Core 7. Install or upgrade your PowerShell version: https://aka.ms/pscore6

RytoEX commented on README.md:
```diff
@@ -1,76 +1,30 @@
-|libx264|[GitHub](https://github.com/mirror/x264/commit/b684ebe04a6f80f8207a57940a1fa00e25274f81)|r3059|
-|ntv2|[GitHub](https://github.com/aja-video/ntv2/commit/abf17cc1e7aadd9f3e4972774a3aba2812c51b75)|16.1|
-|Qt|[Qt.io](https://download.qt.io/official_releases/qt/5.15/5.15.2/single/qt-everywhere-src-5.15.2.tar.xz)|5.15.2|
+OBS dependencies for Windows can be built via the `Build-Dependencies.ps1` PowerShell script. For best compatibility it is advised to use a recent version of PowerShell Core (pwsh). Older versions of PowerShell might work, but support for these is not provided.
```
"For best compatibility it" -> "For best compatibility, it"

RytoEX commented on README.md:
```diff
@@ -1,76 +1,30 @@
-* Enter the `obs-deps` directory
-* Run `bash ./CI/build-deps-macos.sh` to build main dependencies
-* Run `bash ./CI/build-qt-macos.sh` to build Qt dependency
+Further details can be found in the [Wiki Pages](https://github.com/obsproject/obs-deps/wiki)
```
End with punctuation.

RytoEX commented on deps.ffmpeg/50-libtheora.zsh:
```diff
@@ -0,0 +1,163 @@
+local -A versions=(
+  macos 1.1.1
+  linux 1.1.1
+  windows 1.1.1
```
Does this mean we're building libtheora for Windows even if we're not using it in FFmpeg yet? I'd prefer if we just didn't build it for Windows at all until a later commit or PR. You could even stash the changes away and make a separate PR right after this one, and that would be fine with me. To me, it makes more sense that way, and also would provide great examples for how to add functionality to the build scripts.

PatTheMav replied:
It's excluded by the `targets` variable below, the build script ignores any target not contained in it, _if_ it's specified. So in effect the file is entirely skipped right now.

RytoEX replied:
I suppose that's fine. We'll just have to specify the version number for Windows whenever we do enable it.

RytoEX commented on utils.pwsh/Invoke-GitCheckout.ps1:
```diff
@@ -0,0 +1,117 @@
+        .SYNOPSIS
+            Sets a git config value.
+        .DESCRIPTION
+            Allows setting single or multiple config values in a PowerShell-friendly fashion
```
End with punctuation.

RytoEX commented on utils.pwsh/Invoke-GitCheckout.ps1:
```diff
@@ -0,0 +1,117 @@
+        .DESCRIPTION
+            Wraps the git executable with PowerShell syntax to check out
+            a specified Git repository with a given commit hash and branch,
+            or a GitHub pull request id.
```
id -> ID

RytoEX reviewed (changes requested):
A few more items and follow-ups.

PatTheMav commented on deps.ffmpeg/30-svt-av1.zsh:
```diff
@@ -0,0 +1,94 @@
+local name='svt-av1'
+local version='0.9.0'
+local url='https://github.com/AOMediaCodec/SVT-AV1.git'
+local hash='fdf4f36867bba068e221965d12c5975c509845da'
```
Should be downgraded to 0.8.6 again.

RytoEX commented on utils.pwsh/Setup-Host.ps1:
```diff
@@ -0,0 +1,45 @@
+        $script:ProjectRoot = $($script:PSScriptRoot)
+    }
+
+    $script:WorkRoot = "${ProjectRoot}\windows_build_temp"
```
I notice that this uses `windows_build_temp` rather than the current `windows_native_build_temp`. Was this intentional?

PatTheMav replied:
Yes, didn't see a reason to call it specifically the "native" build temp. Every download/dependency is access specifically in that directory, so sharing it between Windows and WSL (if necessary) shouldn't be an issue.

RytoEX replied:
I think I mostly did this originally to avoid clobbering any of the cross-compiled stuff, particularly mbedtls, which I knew we had to build twice, and nasm, which we built manually to set up the cross-compile environment but then also deployed a prebuilt version on Windows (which I will be correcting after this lands).

If you think this is safe enough to do, I won't worry about it here. If we find issues with it later, we can address it later.

RytoEX reviewed (commented):
In addition to my single review comment, I noticed that the dependency update commits have prefixes ("libvpx", "x264", "srt"). Are these what you'd like to use going forward, or are these just a "these are good enough for now" sort of thing?

I noticed that there is no longer a single Windows package similar to the current dependencies2019.zip structure. Is that intentional? Is this something that you plan to address later, or would you like me to address it? Or do you think it's unnecessary? Does the obs-studio CMake rework make it unneeded?

Please get back to me about [my previous questions](https://github.com/obsproject/obs-deps/pull/98#issuecomment-1069698000).

I ran this locally and it seemed that my PowerShell window got stuck on a WebRequest somewhere until I pressed the Enter key. I still need to run full tests using these to build OBS.

PatTheMav commented:
> That is frustrating, yeah. What's the tradeoff with Git checkouts versus file downloads? Data transfer, storage space?

I'd expect that a git checkout always carries an overhead because of the history and object fetches vs downloading the pure unversioned source tarball.

> Would it be easier for us to coordinate on this later after this PR instead of addressing now?

That was my expectation, to just leave it as-is (as we're building it on CI on macOS only) and add Windows in a later change.

> Let me know how this goes or if it's not possible.

The updates have been extracted into separate commits.

> Ah, I more meant if it was mentioned in the PR description and/or commit message. Something like, "Invocation of individual component scripts directly is no longer supported. Individual components can be built by invoking the main build script and specifying the component (examples for macOS, Windows native, and Windows cross)."

Not explicitly. If need be this should be added to the wiki - we haven't had any documentation so far, so that's where I added most information.

RytoEX commented:
Once I can verify that I can build OBS against these and run some tests locally, I'll approve this if I have no other review comments.

tytan652 commented:
I tested building macOS deps with this branch and using it to build OBS Studio. But when running OBS from Xcode many logs about classes being implemented in two binary (OBS and a plugin) and I end up with a SIGABRT while loading `frontend-tools`:
> QWidget: Must construct a QApplication before a QWidget

I built obs-deps by using:
```zsh
./build-deps.zsh
./build-ffmpeg.zsh
./build-qt.zsh
```
And I extracted the three tarball in folder which became the `CMAKE_PREFIX_PATH` of OBS Studio.

PatTheMav commented:
> I tested building macOS deps with this branch and using it to build OBS Studio. But when running OBS from Xcode, I get many logs about classes being implemented in two binary (OBS and a plugin) and I end up with a SIGABRT while loading `frontend-tools`:
> 
> > QWidget: Must construct a QApplication before a QWidget
> 
> I built obs-deps by using:
> 
> ```shell
> ./build-deps.zsh
> ./build-ffmpeg.zsh
> ./build-qt.zsh
> ```
> 
> And I extracted the three tarball in folder which became the `CMAKE_PREFIX_PATH` of OBS Studio.

Did you use a clean build folder or did you ‘update’ an existing one with a new prefix path? Because I’ve been building with those deps fine so far, but always used a ‘fresh’ build folder to avoid it picking up a mix of old and new libs.

tytan652 commented:
I always removed the build folder and regenerate it with a CMake command. Maybe I'm just bad at macOS stuff.

PatTheMav commented:
Tried this on a different Mac which has no prior build environment - fetched OBS recursively, download the three dependencies and copied them into the same folder and just did the following:

* `cmake -S . -B build -G Ninja -DCMAKE_PREFIX_PATH=$(pwd)/../dept`
* `cmake --install build --prefix $(pwd)/build/install`
* `open build/install/OBS.app`

No errors and OBS launches as expected. Only thing you have to ensure is that Qt is extracted correctly (with symlinks in place inside the Frameworks), or else the `codesign` won't be able to identify the bundle type correctly (macOS is somewhat strict about the internal directory and symlink setup of bundles).

tytan652 commented:
The issue was effectively on my side, no Qt libs in the Frameworks folder. Sorry for the false alarm.
I will use regular obs-deps for now.

RytoEX commented:
Following off-thread conversations with @PatTheMav , we've found and addressed a few further issues here. I will be doing some comprehensive testing on Windows next, most likely next week.

RytoEX commented:
It occurs to me now that the deps artifacts have different filename patterns:
* deps-macos
* windows-deps

We might wish for these to follow the same pattern.

RytoEX commented:
I tested this on obsproject/obs-studio@2a0d8d1c9c before today's force push. Here are my results. Some of this was relayed before posting here, so some of this may have already been addressed. Posting here for tracking.

# Good news

OBS builds.

These worked:
* Encoders
  * NVENC
  * QSV
* Recording
* Streaming
  * RTMPS
  * HLS
* Virtual Camera
  * OBS Studio (64-bit)
  * Discord Stable 128323 (5628003) (32-bit)
  * Chrome Version 101.0.4951.67 (Official Build) (64-bit)
  * Firefox 
* Sources
  * Browser
    * Default
    * FishGL/Aquarium
  * Color
  * Display Capture
    * WGC
  * Game Capture
  * Image
    * Animated GIF
    * JPG
    * PNG
  * Image Slide Show
  * Media
    * FLV
    * mp4 with chroma key filter
    * webm with transparency
  * Text (FreeType 2)
  * Text (GDI+)
  * Video Capture Device
  * VLC
    * playlist of files
  * Window Capture
    * BitBlt
    * WGC
 * Filters
   * Chroma Key
   * Color Correction
 * webm stingers

# Bad news
These didn't work:
* x264
  * DLLs are missing from OBS bin dir
  * causes error in Simple Output Mode
  * missing from encoder list in Advanced Output Mode
* Scripting
  * obs-scripting.dll and luajit.dll are in bin dir
  * obs-scripting data exists in data dir
  * Scripting items are missing from the Tools menu

The x264 thing should be fixed by the latest force push here.

RytoEX commented:
Following up on [my previous comment](https://github.com/obsproject/obs-deps/pull/98#issuecomment-1126855675), I can confirm that all of the issues listed there are now resolved.

RytoEX reviewed (approved):
This is _probably_ good to go, unless there are any lingering, unanswered questions from earlier in the thread.

pkviet commented:
tested against my native srt/rist PR. Streams fine to srt/rist servers. Works great.
time to meeeeerge !

PatTheMav commented:
Rebased and force-pushed - I'll let someone else do the honours of merging this. I have a branch with Qt5+Qt6 build support ready once this is merged.

pkviet commented:
I don't have merge rights or I would. Or you can merge symbolically for me.
😛

PatTheMav commented:
> I don't have merge rights or I would. Or you can merge symbolically for me.
> 😛

Nah, let's have Ryan do it. He deserved it for all his review efforts.
